functions/go/set-image/transformer/images.go

```diff
@@ -0,0 +1,237 @@
+		functionConfig.GetOrDie(&imageTrans.Image, "image")
+		functionConfig.GetOrDie(&imageTrans.AdditionalImageFields, "additionalImageFields")
+	default:
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap` with correct formatting",
+			functionConfig.GetKind(), functionConfig.GetAPIVersion(), FnConfigKind)
+	}
+	if err := imageTrans.validateInput(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// validateInput validates the inputs passed into via the functionConfig
+func (imageTrans *ImageTransformer) validateInput() error {
+	if imageTrans.Image.Name == "" && imageTrans.Image.ContainerName == "" && imageTrans.Image.ImageName == "" {
+		return fmt.Errorf("either image name or container name is required")
+	}
+	if imageTrans.Image.NewName == "" && imageTrans.Image.NewTag == "" && imageTrans.Image.Digest == "" {
+		return fmt.Errorf("either image newName, newTag, or digest is required")
+	}
+	if imageTrans.Image.NewTag != "" && imageTrans.Image.Digest != "" {
+		return fmt.Errorf("image newTag and digest both set")
+	}
+	// deprecated name means image name
+	if imageTrans.Image.Name != "" {
+		imageTrans.Image.ImageName = imageTrans.Image.Name
+	}
+	return nil
+}
+
+// Transform updates the image in pod
+func (imageTrans *ImageTransformer) Transform(objects fn.KubeObjects) error {
+	// using unit test and pass in empty string would provide a nil; an empty file in e2e would provide 0 object
+	if objects.Len() == 0 || objects[0] == nil {
+		newResult := fn.GeneralResult("no input resources", fn.Info)
+		imageTrans.Results = append(imageTrans.Results, newResult)
+		return nil
+	}
+	for _, o := range objects.WhereNot(func(o *fn.KubeObject) bool { return o.IsLocalConfig() }) {
+		imageTrans.setPodContainers(o)
+		imageTrans.setPodSpecContainers(o)
+	}
+	return nil
+}
+
+// addWarning provides a warning message when the image name matches correctly while container name does not
+func (imageTrans *ImageTransformer) addWarning(o *fn.KubeObject) {
+	warning := &fn.Result{
+		Message:  "container name does not match, only image name matches",
+		Severity: fn.Warning,
+		ResourceRef: &fn.ResourceRef{
+			APIVersion: o.GetAPIVersion(),
+			Kind:       o.GetKind(),
+			Name:       o.GetName(),
+			Namespace:  o.GetNamespace(),
+		},
+		File: &fn.File{
+			Path:  o.PathAnnotation(),
+			Index: o.IndexAnnotation(),
+		},
+		Tags: nil,
+	}
+	imageTrans.Results = append(imageTrans.Results, warning)
+}
+
+func (imageTrans *ImageTransformer) setPodSpecContainers(o *fn.KubeObject) {
+	if spec := o.GetMap("spec"); spec != nil {
+		if template := spec.GetMap("template"); template != nil {
+			if podSpec := template.GetMap("spec"); podSpec != nil {
+				for _, vecObj := range podSpec.GetSlice("containers") {
+					fieldPath := FieldPath{"spec", "template", "spec", "containers"}
+					imageTrans.updateImages(vecObj, &imageTrans.Image, o, fieldPath)
```

Comment:
Because it is a slice of object, so I have to use `GetSlice` to iterate each subObject. The entire field spec only works when there is no slice involved

Reply:
My suggestion is to not pass fieldpath and logs the result in the caller function